security/engine/engine.go

```diff
@@ -0,0 +1,150 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package engine contains functions/structs for CEL engine.
+package engine
+
+import (
+	"github.com/envoyproxy/go-control-plane/envoy/config/rbac/v2"
+	"google.golang.org/genproto/googleapis/api/expr/v1alpha1"
+	"google.golang.org/grpc/peer"
+)
+
+// Decision enum
+type Decision int32
+
+const (
+	DECISION_ALLOW		Decision = 0
+	DECISION_DENY  		Decision = 1
+	DECISION_UNKNOWN	Decision = 2
+)
+
+var Decision_name = map[int32]string{
+	0: "ALLOW",
+	1: "DENY",
+	2: "UNKNOWN",
+}
+
+var Decision_value = map[string]int32{
+	"ALLOW": 	0,
+	"DENY":  	1,
+	"UNKNOWN": 	2,
+}
+
+// output of CEL engine
+type AuthorizationDecision struct {
+	decision 				Decision
+	authorizationContext	string
+}
+
+// Action enum
+type Action int32
+
+const (
+	ACTION_ALLOW	Action = 0
+	ACTION_DENY  	Action = 1
+)
+
+var Action_name = map[int32]string{
+	0: "ALLOW",
+	1: "DENY",
+}
+
+var Action_value = map[string]int32{
+	"ALLOW": 	0,
+	"DENY":  	1,
+}
+
+// struct for CEL expressions
+type Condition struct {
+	expr	*expr.Expr
+}
+
+func (condition Condition) matches(peerInfo *peer.Peer) bool {
+	// TODO
+	return false
+}
+
+// struct for CEL engine
+type CelEvaluationEngine struct {
+	action		Action
+	conditions	map[string]Condition
+}
+
+// Builds a CEL evaluation engine from Envoy RBAC.
+func RbacToCelEvaluationEngine(rbac envoy_config_rbac_v2.RBAC) CelEvaluationEngine {
+	var action Action
+	if rbac.Action == envoy_config_rbac_v2.RBAC_ALLOW {
+		action = ACTION_ALLOW
+	} else {
+		action = ACTION_DENY
+	}
+	
+	conditions := make(map[string]Condition)
+	for policyName, policy := range rbac.Policies {
+        conditions[policyName] = Condition{policy.Condition}
+	}
+	
+	return CelEvaluationEngine{action, conditions}
+}
+
+// Builds a CEL evaluation engine from runtime policy template.
+// Currently do not have access to RTPolicyTemplate/can't import
+// func RTPolicyTemplateToCelEvaluationEngine(rt_policy RTPolicyTemplate) CelEvaluationEngine {
+// 	// TODO
+// 	return nil
+// }
+
+// The core function that evaluates whether an RPC is authorized
+func (engine CelEvaluationEngine) evaluate(peerInfo *peer.Peer) AuthorizationDecision {
+	for policyName, condition := range engine.conditions {
+		if condition.matches(peerInfo) {
+			var decision Decision
+			if engine.action == ACTION_ALLOW {
+				decision = DECISION_ALLOW
+			} else {
+				decision = DECISION_DENY
+			}
+			return AuthorizationDecision{decision, "Policy matched: " + policyName}
+		}
+	}
+	// if no conditions matched
+	return AuthorizationDecision{DECISION_UNKNOWN, "No policies matched"}
+}
+
+// list of engines to evaluate against
```

Comment:
You don't have to worry about the engine chaining - for now let's just assume there will always be only one engine. This will also create a package level variable, which is not ideal.